net/host.go

```diff
@@ -308,86 +334,113 @@ func (h *host) addrInfo() peer.AddrInfo {
+// io.ErrUnexpectedEOF is returned.
+func readStreamMessage(s io.Reader) (Message, error) {
+	if s == nil {
+		return nil, errNilStream
 	}
 
-	var (
-		tot int
-	)
-
-	length, bytesRead, err := readLEB128ToUint64(stream, buf[:1])
+	lenBuf := make([]byte, 4) // uint32 size
+	n, err := io.ReadFull(s, lenBuf)
 	if err != nil {
-		return bytesRead, fmt.Errorf("failed to read length: %w", err)
-	}
-
-	if length == 0 {
-		return 0, nil
+		if isEOF(err) {
+			if n > 0 {
+				err = io.ErrUnexpectedEOF
+			} else {
+				err = io.EOF
+			}
+		}
+		return nil, err
 	}
+	msgLen := binary.LittleEndian.Uint32(lenBuf)
 
-	if length > uint64(len(buf)) {
-		log.Warnf("received message with size greater than allocated message buffer: msg size=%d, buffer size=%d",
-			length, len(buf))
-		return 0, fmt.Errorf("message size greater than allocated message buffer: got %d", length)
+	if msgLen > maxMessageSize {
+		log.Warnf("Received message longer than max allowed size: msg size=%d, max=%d",
+			msgLen, maxMessageSize)
+		return nil, fmt.Errorf("message size %d too large", msgLen)
 	}
 
-	tot = 0
-	for i := 0; i < maxReads; i++ {
-		n, err := stream.Read(buf[tot:])
-		if err != nil {
-			return n + tot, err
-		}
-
-		tot += n
-		if tot == int(length) {
-			break
+	msgBuf := make([]byte, msgLen)
+	_, err = io.ReadFull(s, msgBuf)
+	if err != nil {
+		if isEOF(err) {
+			err = io.ErrUnexpectedEOF
 		}
+		return nil, err
 	}
 
-	if tot != int(length) {
-		return tot, fmt.Errorf("failed to read entire message: expected %d bytes, received %d bytes", length, tot)
-	}
+	return message.DecodeMessage(msgBuf)
+}
 
-	return tot, nil
+func isEOF(err error) bool {
+	switch {
+	case
+		errors.Is(err, net.ErrClosed), // what libp2p with QUIC usually generates
+		errors.Is(err, io.EOF),
+		errors.Is(err, io.ErrUnexpectedEOF),
+		errors.Is(err, io.ErrClosedPipe):
+		return true
+	default:
+		return false
+	}
 }
 
 // bootstrap connects the host to the configured bootnodes
 func (h *host) bootstrap() error {
-	failed := 0
-	for _, addrInfo := range h.bootnodes {
-		log.Debugf("bootstrapping to peer: peer=%s", addrInfo.ID)
-		err := h.h.Connect(h.ctx, addrInfo)
-		if err != nil {
-			log.Debugf("failed to bootstrap to peer: err=%s", err)
-			failed++
+
+	if len(h.bootnodes) == 0 {
+		log.Warnf("Bootstraping peers skipped, no bootnodes found")
+		return nil
+	}
+
+	selfID := h.PeerID()
+
+	var failed uint64 = 0
+	var wg sync.WaitGroup
+	for _, bn := range h.bootnodes {
+		if bn.ID == selfID {
+			continue
 		}
+		h.h.Peerstore().AddAddrs(bn.ID, bn.Addrs, peerstore.PermanentAddrTTL)
+		log.Debugf("Bootstrapping to peer: %s (%s)", bn, h.h.Network().Connectedness(bn.ID))
+		wg.Add(1)
+		go func(p peer.AddrInfo) {
+			defer wg.Done()
+			err := h.h.Connect(h.ctx, p)
+			if err != nil {
+				log.Debugf("Failed to bootstrap to peer %s: err=%s", p.ID, err)
+				atomic.AddUint64(&failed, 1)
+			}
+			log.Debugf("Bootstrapped connections to: %s", h.h.Network().ConnsToPeer(p.ID))
```

Comment:
what does this log actually look like? seems like printing a `Conn` type might be ugly?

Reply:
It definitely is ugly and maybe we want to remove it long term. On the flip side, to get the less ugly pieces of information, I'd have to do a lot of ugly casting here.
```
Bootstrapped connections to: [<swarm.Conn[*libp2pquic.transport] /ip4/0.0.0.0/udp/9900/quic-v1 (12D3KooWQQWDJ7KA1Fwdf2ejWz9VXHKvY8cC5PB7Sf34fbEGbsgV) <-> /ip4/149.28.254.150/udp/9900/quic-v1 (12D3KooWGHwHJt6uYDC8fJaeGrduDYwjMBG86Lv1XmopnygwJBzN)>]
```

Reply:
I figured out how to clean it up.  It looks better now:
```
Bootstrapped connection to /ip4/164.92.217.101/udp/9900/quic-v1/p2p/12D3KooWQLZjDoEtjuVJs2DFsFbfasdxpJpJ5Q7jVztoDGzUa3Qz
```

Reply:
nice :D